pulser-core/pulser/math/abstract_array.py

```diff
@@ -0,0 +1,301 @@
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines the AbstractArray class."""
+from __future__ import annotations
+
+import functools
+import importlib.util
+import operator
+from typing import Any, Generator, Union, cast
+
+import numpy as np
+from numpy.typing import ArrayLike, DTypeLike
+
+from pulser.json.utils import obj_to_dict
+
+try:
+    import torch
+except ImportError:  # pragma: no cover
+    pass
+
+
+class AbstractArray:
+    """An abstract array containing an array or tensor.
+
+    Args:
+        array: The array to store.
+    """
+
+    def __init__(
+        self,
+        array: AbstractArrayLike,
+        dtype: DTypeLike = None,
+        force_array: bool = False,
+    ):
+        """Initializes a new AbstractArray."""
+        self._array: np.ndarray | torch.Tensor
+        if isinstance(array, AbstractArray):
+            self._array = array._array
+        elif self.has_torch() and isinstance(array, torch.Tensor):
+            self._array = torch.as_tensor(
+                array,
+                dtype=dtype,  # type: ignore[arg-type]
+            )
+        else:
+            self._array = np.asarray(array, dtype=dtype)
+
+        if force_array and self._array.ndim == 0:
+            self._array = self._array[None]
+
+    @staticmethod
+    @functools.lru_cache
+    def has_torch() -> bool:
+        """Checks whether torch is installed."""
+        return importlib.util.find_spec("torch") is not None
+
+    @functools.cached_property
+    def is_tensor(self) -> bool:
+        """Whether the stored array is a tensor."""
+        return self.has_torch() and isinstance(self._array, torch.Tensor)
+
+    def astype(self, dtype: DTypeLike) -> AbstractArray:
+        """Casts the data type of the array contents."""
+        if self.is_tensor:
+            return AbstractArray(
+                cast(torch.Tensor, self._array).to(
+                    dtype=dtype  # type: ignore[arg-type]
+                )
+            )
+        return AbstractArray(cast(np.ndarray, self._array).astype(dtype))
+
+    def as_tensor(self) -> torch.Tensor:
+        """Converts the stored array to a torch Tensor."""
+        if not self.has_torch():
+            raise RuntimeError("`torch` is not installed.")
+        return torch.as_tensor(self._array)
+
+    def as_array(self, *, detach: bool = False) -> np.ndarray:
+        """Converts the stored array to a Numpy array.
+
+        Args:
+            detach: Whether to detach before converting.
+        """
+        if detach and self.is_tensor:
+            return cast(torch.Tensor, self._array).detach().numpy()
```

Comment:
Keep in mind that `AbstractArray` is not supposed to be a user facing object, we want the users to keep providing either numpy arrays or torch tensors. Rather, `AbstractArray` allows us to internally handle both input types adequately.

> Perhaps we could also have the `detach()` method implemented ?

I have not had the need to have a `AbstractArray.detach()` method so far, nor do I see how it could be useful tbh.

> I think you (almost) always have to detach() a torch tensor to be able to convert it to a numpy array. Additionnaly, having the `detach` would also enable the serializing of a torch tensor (detaching it from its graph) https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html#torch.Tensor.detach

Not really, you only have to `detach()` when `requires_grad=True`. That is way this method allows you to choose whether to detach or not: if you don't care about the computational graph, you can set `detach=True`; otherwise it won't detach so as to preserve the computational graph and will eventually fail if you are not handling things properly. 

> More globally, I think I understand from this function that you don't want to implement all the methods a torch tensor and a numpy array would have in their syntax, but rather develop a new syntax to implement all these methods. It's fine by me, but I am wondering if it's good from a user point of view, if we want to think of a user programming in the torch framework with other purposes than differentiation (as suggested by @awennersteen) ?

Again, keeping in mind that `AbstractArray` should only be used internally, we only need to implement the methods and functions we need, as that need arises. In the few instances where a user is given an `AbstractArray`, we expect them to convert them to the type of their choice (`array` or `tensor`) before doing whatever computations they want. Nonetheless, you should be able to treat an `AbstractArray` as a `np.darray` the majority of the time.